File: src/main/java/com/hoi4utils/clausewitz/localization/Localization.java
package com.hoi4utils.clausewitz.localization;

import com.hoi4utils.clausewitz.exceptions.UnexpectedLocalizationStatusException;

import java.io.File;

/**
 * This is the Localization record file.
 * // todo add explicit/default loc key
 */
public record Localization(String ID, Integer version, String text, Status status) {

	public Localization(String id, String text, Status status) {
		this(id, null, text, status);
	}

	/**
	 * Creates a new localization with the given ID, and text. version optional
	 */
	public Localization(String key, String text) {
		this(key, null, text, Status.NEW);
	}

	public Localization(String key, String text, Integer version) {
		this(key, version, text, Status.NEW);
	}


	public enum Status {
//		DEFAULT,
		EXISTS, // localization already exists in a mod localization file (was only read into the
				// program)
		NEW,    // localization is new (not written to a loc file). If new localization is saved, it's ID should be unique.
		UPDATED, // localization key already exists in a loc file, but text is updated. This localization can overwrite the existing localization.
		VANILLA, // localization is from the base ('vanilla') game
		MISSING, // localization is completely missing (no localization key exists in the localization list)
	}

	public boolean isReplaceableBy(Localization other) throws UnexpectedLocalizationStatusException {
		switch (status) {
			case NEW:
				if (other.status != Status.NEW) throw new UnexpectedLocalizationStatusException(this, other);
				else return true;
			case EXISTS:
				if (other.status != Status.UPDATED) throw new UnexpectedLocalizationStatusException(this, other);
				else return true;
			case UPDATED:
				if (other.status != Status.UPDATED && other.status != Status.EXISTS) {
					throw new UnexpectedLocalizationStatusException(this, other);
				}
				else return true;
			case VANILLA:
				return false;
			default:
				return false;
		}
	}

	public boolean isReplaceable() {
		return switch (status) {
			case NEW -> true;
			case EXISTS -> true;
			case UPDATED -> true;
			case VANILLA -> false;
			default -> false;
		};
	}

	public boolean isNew() {
		return status == Status.NEW;
	}

	/**
	 * Replaces the text of this localization with the given text.
	 * @param text
	 * @return a new localization with the given text.
	 */
	public Localization replaceWith(String text) {
		return new Localization(ID, version, text, Status.UPDATED);
	}

	public Localization replaceWith(String text, Integer version, File file) {
		var status = switch (this.status) {
			case NEW -> Status.NEW;
			case EXISTS -> Status.UPDATED;
			case UPDATED -> Status.UPDATED;
			default -> throw new UnexpectedLocalizationStatusException(this);
		};
		return new Localization(ID, version, text, status);
	}

	public String toString() {
		return ID + ":" + (version == null? "" : version) + " " + "\"" + text + "\"";
	}

	public String baseKey() {
		if (ID.endsWith("_desc")) {
			return ID.substring(0, ID.length() - 5);
		} else return ID; 
	}

}

File: src/main/scala/com/hoi4utils/localization/Localization.scala
package com.hoi4utils.localization

import com.hoi4utils.exceptions.UnexpectedLocalizationStatusException
import com.hoi4utils.localization.Localization.Status

import java.io.File

case class Localization(
                         ID: String,
                         version: Option[Int],
                         text: String,
                         status: Localization.Status
                       ) {

  /** Does this record allow itself to be replaced by `other`? */
  @throws[UnexpectedLocalizationStatusException]
  def isReplaceableBy(other: Localization): Boolean = status match {
    case Status.NEW =>
      if other.status != Status.NEW then
        throw UnexpectedLocalizationStatusException(this, other)
      else true

    case Status.EXISTS =>
      if other.status != Status.UPDATED then
        throw UnexpectedLocalizationStatusException(this, other)
      else true

    case Status.UPDATED =>
      if other.status != Status.UPDATED && other.status != Status.EXISTS then
        throw UnexpectedLocalizationStatusException(this, other)
      else true

    case Status.VANILLA => false
    case Status.MISSING => false
  }

  /** Can this localization be replaced by *any* new one? */
  def isReplaceable: Boolean = status match
    case Status.NEW | Status.EXISTS | Status.UPDATED => true
    case Status.VANILLA | Status.MISSING               => false

  def isNew: Boolean = status == Status.NEW

  /** Copy with new text; always bumps you to UPDATED. */
  def replaceWith(newText: String): Localization =
    copy(text = newText, status = Status.UPDATED)

  /**
   * Copy with new text and version, preserving NEW→NEW but EXISTS→UPDATED, etc.
   */
  @throws[UnexpectedLocalizationStatusException]
  def replaceWith(newText: String, newVersion: Option[Int], file: File): Localization = {
    val newStatus = status match
      case Status.NEW     => Status.NEW
      case Status.EXISTS  => Status.UPDATED
      case Status.UPDATED => Status.UPDATED
      case _               => throw UnexpectedLocalizationStatusException(this)
    copy(version = newVersion, text = newText, status = newStatus)
  }

  override def toString: String =
    s"$ID:${version.map(_.toString).getOrElse("")} \"$text\""

  /** If your key ends in "_desc", strip that suffix for the base key. */
  def baseKey: String =
    if ID.endsWith("_desc") then ID.stripSuffix("_desc") else ID
}

object Localization {
  enum Status {
    case EXISTS, NEW, UPDATED, VANILLA, MISSING
  }

  /** Java‐style overloads of your record‐constructors: */
  def apply(id: String, text: String, status: Status): Localization =
    Localization(id, None, text, status)

  def apply(id: String, text: String): Localization =
    Localization(id, None, text, Status.NEW)

  def apply(id: String, text: String, version: Int): Localization =
    Localization(id, Some(version), text, Status.NEW)
}
